### PR TITLE
COMMERCE-11163 Supplier Role to manage their Orders

### PR DESCRIPTION
**Pull request by brianikim**
Resent from: https://github.com/liferay-commerce/liferay-portal/pull/3870

@luca-pellizzon ,

https://issues.liferay.com/browse/COMMERCE-11163

### CODE REVIEW

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/permission/CommerceChannelPermissionImpl.java [hunk 1]
```diff
@@ -133,7 +150,53 @@ private boolean _contains(
 			commerceChannel.getCommerceChannelId(), actionId);
 	}
 
+	private boolean _hasSupplierAccount(
```
Comment (luca-pellizzon) [inline modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/permission/CommerceChannelPermissionImpl.java:153]: pls check how this stack up with https://github.com/liferay-commerce/liferay-portal/pull/3917

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java [hunk 2]
```diff
@@ -220,6 +223,36 @@ else if (name.equals(AccountRoleConstants.ROLE_NAME_SUPPLIER)) {
 					new String[] {ActionKeys.ACCESS_IN_CONTROL_PANEL});
 			}
 
+			companyResourceActionIds.put(
+				"com.liferay.commerce.inventory",
+				new String[] {CommerceInventoryActionKeys.MANAGE_INVENTORY});
```
Comment (luca-pellizzon) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java:228]: Supplier should not have MANAGE_INVENTORY permission, it's too generic and it will cause other issues in the inventory section.
Please check what it is used for and find another solution

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java [hunk 3]
```diff
@@ -91,7 +100,9 @@ public boolean contains(
 		throws PortalException {
 
 		if (_accountEntryModelResourcePermission.contains(
-				permissionChecker, accountEntryId, actionId)) {
+				permissionChecker, accountEntryId, actionId) ||
```
Comment (luca-pellizzon) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java:103]: Supplier should not access all accounts, only the one he is member of.
Is this needed for showing order's details?

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java [hunk 4]
```diff
@@ -240,7 +240,8 @@ else if (name.equals(AccountRoleConstants.ROLE_NAME_SUPPLIER)) {
 					"MANAGE_COMMERCE_ORDER_PAYMENT_TERMS",
 					"MANAGE_COMMERCE_ORDER_PRICES",
 					"MANAGE_COMMERCE_ORDER_RESTRICTED_NOTES",
-					"MANAGE_COMMERCE_ORDER_SHIPPING_OPTIONS"
+					"MANAGE_COMMERCE_ORDER_SHIPPING_OPTIONS",
+					"MANAGE_COMMERCE_ORDERS"
```
Comment (luca-pellizzon) [inline modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java:244]: MANAGE_COMMERCE_ORDERS why is this needed? can't we manage it with the custom logic instead? This is givin him access to manage any order

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/permission/CommerceChannelPermissionImpl.java (hunk 1 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/permission/CommerceChannelPermissionImpl.java:153]: #3917 needs to be updated to prevent the duplicate. The commits for COMMERCE-11163 in that PR were done using old commits. I left a comment on the PR.

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java (hunk 3 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java:103]: Hi @luca-pellizzon , yes this is needed. The alternative is to change this method permission check: https://github.com/liferay-commerce/liferay-portal/blob/42246a7dccec49db7a18d3601dd299dd4b083974/modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemServiceImpl.java#L294

Comment (brianikim) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java:103]: Used in: https://github.com/liferay-commerce/liferay-portal/blob/42246a7dccec49db7a18d3601dd299dd4b083974/modules/apps/commerce/commerce-shipment-web/src/main/java/com/liferay/commerce/shipment/web/internal/frontend/data/set/provider/CommerceShippableOrderItemsFDSDataProvider.java#L81

Comment (luca-pellizzon) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java:103]: I guess we should use the localservice in this fds, you should be able to access orderItems as long as you can access the related order. We will fix the call later. Also I don't get why we filter by account there

Comment (brianikim) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java:103]: I will take care of this with a `localService` call instead.

File: modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/permission/CommerceChannelPermissionImpl.java (hunk 1 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-product-service/src/main/java/com/liferay/commerce/product/internal/permission/CommerceChannelPermissionImpl.java:153]: I communicated with @dsitu  on this. @luca-pellizzon , we should be able to resolve the commit duplicate/dependencies soon.

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by brianikim** [comment]
ci:test:relevant

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java (hunk 3 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/security/permission/resource/AccountEntryModelResourcePermission.java]: Address issues fixed here: https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/3f7a41c86f21e3bf50e28cffb7cf1b08b58a0fd6

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java (hunk 4 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java]: Opted to use localService instead here: https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/77f4ce8b5e1a4b9db76f9477f54b10a8817a5eb5

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java (hunk 2 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java]: Opted to use localService instead here: https://github.com/liferay-commerce/liferay-portal/commit/77f4ce8b5e1a4b9db76f9477f54b10a8817a5eb5

File: modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java (hunk 4 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/internal/util/CommerceAccountRoleHelperImpl.java]: In the shipment view, the permission check is done on the CommerceOrder of the address (This will always be the parent commerecOrder, the original non-split) which Supplier does not have permission to. Opted to use localService instead for all address related fetch calls in shipment-web: https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/3f7a41c86f21e3bf50e28cffb7cf1b08b58a0fd6

**Comment by brianikim** [comment]
Test failures due to documented QA failure on the CI side: https://github.com/brianchandotcom/liferay-portal/commit/10decbf7f21364f7611fd25c309cd2ed855fabdb. Need this commit to be in the upstream before :relevant and :commerce can be run again. Let me know if the PR needs any more changes @luca-pellizzon

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce

**Comment by brianikim** [comment]
@luca-pellizzon , I talked to core team. They said no on the permissionChecker implementation. I opted to introduce `secure` parameter instead after discussing with @dsitu . We cannot do a hard change to `localService` because it will allow non-permission checked changes.

**Comment by brianikim** [comment]
ci:stop

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

File: modules/apps/commerce/commerce-api/src/main/java/com/liferay/commerce/order/engine/CommerceOrderEngine.java [hunk 5]
```diff
@@ -47,6 +51,11 @@ public CommerceOrder transitionCommerceOrder(
 			CommerceOrder commerceOrder, int orderStatus, long userId)
 		throws PortalException;
 
+	public CommerceOrder transitionCommerceOrder(
```
Comment (luca-pellizzon) [inline modules/apps/commerce/commerce-api/src/main/java/com/liferay/commerce/order/engine/CommerceOrderEngine.java:54]: don't create new methods, replace the old one and just put the secure value true as default

**Review by luca-pellizzon — CHANGES_REQUESTED** [review]

File: modules/apps/commerce/commerce-api/src/main/java/com/liferay/commerce/order/engine/CommerceOrderEngine.java (hunk 5 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-api/src/main/java/com/liferay/commerce/order/engine/CommerceOrderEngine.java]: Updated PR before I go to sleep @luca-pellizzon .

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Review by luca-pellizzon — APPROVED** [review]

**Comment by lorenzolapescara** [comment]
Manual QA Passed

**Comment by lorenzolapescara** [comment]
ci:test:relevant

**Comment by lorenzolapescara** [comment]
ci:test:commerce-functional

File: modules/apps/commerce/commerce-shipment-web/src/main/java/com/liferay/commerce/shipment/web/internal/portlet/action/EditCommerceShipmentItemMVCActionCommand.java [hunk 6]
```diff
@@ -228,9 +228,10 @@ private CommerceShipmentItem _updateCommerceShipmentItem(
 		CommerceShipmentItem commerceShipmentItem = null;
 
 		List<CommerceInventoryWarehouse> commerceInventoryWarehouses =
-			_commerceInventoryWarehouseService.getCommerceInventoryWarehouses(
-				commerceOrderItem.getCompanyId(),
-				commerceOrderItem.getGroupId(), true);
+			_commerceInventoryWarehouseLocalService.
+				getCommerceInventoryWarehouses(
```
Comment (luca-pellizzon) [inline modules/apps/commerce/commerce-shipment-web/src/main/java/com/liferay/commerce/shipment/web/internal/portlet/action/EditCommerceShipmentItemMVCActionCommand.java:232]: it shouldn't be need to use local service after https://github.com/liferay-commerce/liferay-portal/pull/3933

**Comment by luca-pellizzon** [comment]
@brianikim in here
com/liferay/commerce/shipment/web/internal/frontend/data/set/provider/CommerceShipmentItemFDSDataProvider.java:78
if the supplier can't access the warehouse we get a permission exception.
We should manage it, with like a fetch, and if null we just not going to print the warehouse name.

File: modules/apps/commerce/commerce-shipment-web/src/main/java/com/liferay/commerce/shipment/web/internal/portlet/action/EditCommerceShipmentItemMVCActionCommand.java (hunk 6 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-shipment-web/src/main/java/com/liferay/commerce/shipment/web/internal/portlet/action/EditCommerceShipmentItemMVCActionCommand.java]: https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/ce2a08bef97a89e8e169889092fa8495692cf14f

**Comment by brianikim** [comment]
@luca-pellizzon , see https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/ce2a08bef97a89e8e169889092fa8495692cf14f. I also removed the warehouse permission checks as requested.

**Comment by brianikim** [comment]
@lorenzolapescara , if this needs to be re-tested, commits from latest COMMERCE-11160 and COMMERCE-11699 need ot be included. Otherwise it can be continued.

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by luca-pellizzon** [comment]
@brianikim [CommerceInventoryWarehousePermissionImpl.java](https://github.com/liferay-commerce/liferay-portal/pull/3933/commits/0c5629c2f765b627f34d37d450169f2c2a4ba964#diff-cb77bfcc8f620e13e4d108ebd2be2a6d742d37efdb1bb0c52c56f0a4d8339f83) is implemented.
Even if it was true, we can't let supplier access all warehouses. He should only access those that are linked to his channel.

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by brianikim** [comment]
Hi @luca-pellizzon , I discussed with @CrescenzoRegaLF  on commit: https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/e988319e204f4d8e31462761709fdc943b20e703

We concluded that the localService call is necessary because we do not have access to commerceInventoryWarehouseId until we call:

```
		List<CommerceInventoryWarehouse> commerceInventoryWarehouses =
			_commerceInventoryWarehouseLocalService.
				getCommerceInventoryWarehouses(
					commerceOrderItem.getCompanyId(),
					commerceOrderItem.getGroupId(), true);
```

It is okay we use localService to fetch the commerceInventoryWarehouses because we are going to do modelResourcePermission checks on the returned warehouses and only allow viewing warehouseItems, and adding/updating shipmentItems if it passes the modelResourcePermission check.

@lorenzolapescara , behavior should be the same as you've tested previously. Let me know if you plan on testing this again.

Sincerely,
Brian I. Kim

**Comment by luca-pellizzon** [comment]
@brianikim I don't like the proposed solution. Following the logic of `CommerceInventoryWarehouseItemFDSDataProvider.getItems(...)` we can see that we are retrieving warehouses starting from: `companyId, groupId, active`.
In this case the groupId is the channelGroupId coming from the orderItem and this means that the supplier can access (or not) all the results of that "query". Cycling through them is worthless as we either get all or nothing.
The check we could do inside `CommerceInventoryWarehouseItemFDSDataProvider` is: permission on channel starting from commerceOrderItem.getGroupId().
Also if we wanted to add the model permission check as you are doing it should be done in the "remote" service instead, not in a `FDSDataProvider`.

**Comment by brianikim** [comment]
@luca-pellizzon , I believe the last three commits should satisfy your request.
https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/42870b6a691ffbe186290763628ade61d589e848 New commit to properly handle count.
https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/889f4e32c8f8a55e6b59b5186376607d555c1f5d New commit to handle permissions. Use channelModelResourcePermission if channel exists, otherwise default back to original behavior of checking MANAGE_INVENTORY
https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/69bca209ba82e9336b5d753a4125ad1713e3bcec New commit to handle view. Do not include the warehouse if warehouseItem does not exist.

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by brianikim** [comment]
@luca-pellizzon , see newly updated https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/6246f9693ddd24cbd7077d41354031d4ec8a328a

**Comment by brianikim** [comment]
ci:test:sf

**Review by luca-pellizzon — APPROVED** [review]

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by brianikim** [comment]
@luca-pellizzon ,

I updated https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/0185d372cbc7c41fbd4fe81e27e32e9fbfe9448b

I forgot we needed to do the same logic in [EditCommerceShipmentItemMVCActionCommand.java](https://github.com/liferay-commerce/liferay-portal/pull/3912/commits/0185d372cbc7c41fbd4fe81e27e32e9fbfe9448b#diff-1280bc0d594e7a6d9e93a91bb2a920ac789e5f70349d5c2c43f07b9ad2be2211)

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by lorenzolapescara** [comment]
Manual test passed.

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:commerce-functional

**Comment by CrescenzoRegaLF** [comment]
ci:reevaluate:1257401_1651

**Comment by CrescenzoRegaLF** [comment]
ci:reevaluate:1307401_7003

**Comment by CrescenzoRegaLF** [comment]
ci:test:relevant

**Comment by CrescenzoRegaLF** [comment]
ci:test:commerce-functional

**Comment by victoroscuro** [comment]
ci:stop:commerce-functional

**Comment by lorenzolapescara** [comment]
Even if the relevant run is failing on CI, the previous run's failures are not related to the PR.

**Comment by victoroscuro** [comment]
ci:stop
